File: yoshimoto/temperature_hist.py
import matplotlib.pyplot as plt
import numpy as np
from scipy.stats import norm
import my_style


def temperature_hist():
    nbin = 10
    bin_min = 36.05
    bin_max = 38.05
    bin_width = (bin_max - bin_min) / nbin
    vy = []
    with open("temperature.txt") as f:
        for line in f:
            vy.append(float(line.split()[0]))

    N = len(vy)
    mean = np.mean(vy)
    stdev = np.std(vy)

    label="Entries {0}\nMean {1:.2f}\nStd Dev {2:.4f}".format(N,mean,stdev)
    content = plt.hist(vy, bins=nbin, range=[
                       bin_min, bin_max], histtype="step", edgecolor='k',label=label)


    xx = np.linspace(bin_min, bin_max, 500)
    plt.plot(xx, norm.pdf(xx, mean, stdev)*N*bin_width, color="red")
    plt.xlim(bin_min, bin_max)
    plt.xlabel("Body temperature[$^\\circ$C]")
    plt.ylabel("Frequency")

    chi2 = 0
    for i in range(nbin):
        observed = content[0][i]
        center = (content[1][i]+content[1][i+1])*0.5
        if observed == 0:
            continue
        expected = norm.pdf(center, mean, stdev)*(N*bin_width)

        print("{:.2f} {:.4f} {:.0f}".format(center, expected, observed))
        chi2 += (observed - expected) * (observed - expected) / expected

    print("N    :{}".format(N))
    print("mean :{:.4f}".format(mean))
    print("stdev:{:.4f}".format(stdev))
    print("chi2 :{:.4f}".format(chi2))

    plt.legend()
    plt.show()


if __name__ == "__main__":
    temperature_hist()

File: yoshimoto/temperature_hist.cpp
#include <TH1D.h>
#include <TF1.h>
#include <TMath.h>
#include <fstream>
#include <iostream>
#include "my_style.hpp"
using namespace std;

void temperature_hist()
{
    SetMyStyle();
    double temp;
    int nbin = 10;
    double bin_min = 36.05;
    double bin_max = 38.05;
    double bin_width = (bin_max - bin_min) / nbin;
    TH1D *histgram = new TH1D("temp", "", nbin, bin_min, bin_max);
    ifstream ifs("temperature.txt");
    while (ifs >> temp)
    {
        histgram->Fill(temp);
    }
    histgram->SetTitle(";Body temperature [#circ C];Frequency");
    histgram->Draw();

    int N = histgram->GetEntries();
    double mean = histgram->GetMean();
    double stdev = histgram->GetStdDev();

    TF1 *pdf = new TF1("pdf", "ROOT::Math::gaussian_pdf(x,[0],[1])*[2]*[3]", bin_min, bin_max);
    pdf->SetParameter(0, stdev);
    pdf->SetParameter(1, mean);
    pdf->SetParameter(2, N);
    pdf->SetParameter(3, bin_width);
    pdf->SetLineColor(kRed);
    pdf->SetNpx(500);
    pdf->Draw("same");

    double chi2 = 0;
    for (int i = 1; i <= nbin; i++)
    {
        double observed = histgram->GetBinContent(i);
        if (observed == 0)
        {
            continue;
        }
        double expected = TMath::Gaus(histgram->GetBinCenter(i), mean, stdev, true) * (N * bin_width);
        cout << histgram->GetBinCenter(i) << " " << expected << " " << observed << endl;

        chi2 += (observed - expected) * (observed - expected) / expected;
    }
    cout << "N    :" << N << endl;
    cout << "mean :" << mean << endl;
    cout << "stdev:" << stdev << endl;
    cout << "chi2 :" << chi2 << endl;
}